pkg/daemon/multus/resources.go

```diff
@@ -0,0 +1,279 @@
+	if err != nil {
+		return fmt.Errorf("failed to generate web server config: %w", err)
+	}
+	configMap.SetOwnerReferences(owners) // set owner refs so cleanup is easier
+
+	// create configmap before pod so pod doesn't crashloopbackoff on first creation
+	_, err = vt.Clientset.CoreV1().ConfigMaps(vt.Namespace).Create(ctx, configMap, meta.CreateOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to create web server config: %w", err)
+	}
+
+	_, err = vt.Clientset.CoreV1().Pods(vt.Namespace).Create(ctx, pod, meta.CreateOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to create web server pod: %w", err)
+	}
+
+	return nil
+}
+
+func (vt *ValidationTest) getWebServerInfo(
+	ctx context.Context,
+	desiredPublicNet, desiredClusterNet *types.NamespacedName,
+) (podNetworkInfo, []string, error) {
+	podInfo := podNetworkInfo{}
+
+	pod, err := vt.Clientset.CoreV1().Pods(vt.Namespace).Get(ctx, webServerPodName(), meta.GetOptions{})
+	if err != nil {
+		return podInfo, []string{}, fmt.Errorf("unexpected error when getting web server pod: %w", err)
+	}
+
+	var publicAddr, clusterAddr string
+	publicAddr, clusterAddr, networkSuggestions, err := getNetworksFromPod(pod, desiredPublicNet, desiredClusterNet)
+	if err != nil {
+		return podInfo, networkSuggestions, fmt.Errorf("no web server network info: %w", err)
+	}
+
+	if !podIsReady(*pod) {
+		return podInfo, []string{}, fmt.Errorf("web server pod is not ready yet")
+	}
+
+	podInfo.nodeName = pod.Spec.NodeName
+	podInfo.publicAddr = publicAddr
+	podInfo.clusterAddr = clusterAddr
+	return podInfo, []string{}, nil // no suggestions if successful
+}
+
+func (vt *ValidationTest) startClients(
+	ctx context.Context,
+	owners []meta.OwnerReference,
+	serverPublicAddr, serverClusterAddr string,
+) error {
+	for i := 0; i < vt.DaemonsPerNode; i++ {
+		ds, err := vt.generateClientDaemonSet(i, serverPublicAddr, serverClusterAddr)
+		if err != nil {
+			return fmt.Errorf("failed to generate client daemonset for client #%d: %w", i, err)
+		}
+		ds.SetOwnerReferences(owners) // set owner refs so cleanup is easier
+
+		_, err = vt.Clientset.AppsV1().DaemonSets(vt.Namespace).Create(ctx, ds, meta.CreateOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to create client daemonset for client #%d: %w", i, err)
+		}
+	}
+
+	return nil
+}
+
+func (vt *ValidationTest) getExpectedNumberOfClientPods(ctx context.Context) (int, error) {
+	var agreedNumberScheduled int32
+	listOpts := meta.ListOptions{
+		LabelSelector: clientAppLabel(),
+	}
+	dsets, err := vt.Clientset.AppsV1().DaemonSets(vt.Namespace).List(ctx, listOpts)
+	if err != nil {
+		return 0, fmt.Errorf("unexpected error listing client daemonsets: %w", err)
+	}
+	if len(dsets.Items) < vt.DaemonsPerNode {
+		return 0, fmt.Errorf("got fewer client daemonsets [%d] than should exist [%d]", len(dsets.Items), vt.DaemonsPerNode)
+	}
+
+	for _, d := range dsets.Items {
+		numScheduled := d.Status.CurrentNumberScheduled
```

Comment:
Before this is called, is there some wait period or how do we know if the daemonset pods are done scheduling?

Reply:
I believe the wait logic is sufficient, though complex. But your question gives me enough of a wonder about race conditions that I think adding a wait would be a good idea. 


I did a lot of experimentation to get to the point it is now, and I'll leave my typed out dissection of it in case it's useful when revisiting this months from now.

The first thing this method does is get all the daemonsets. Because we know how many daemonsets we started, we wait in the loop (outside the method) to be sure all daemonsets themselves are scheduled. That is the first part of the wait you are asking about. 

The other part of "done scheduling" actually comes after the lines you flagged, and the whole loop is necessary for context.

```go
	for _, d := range dsets.Items {
		numScheduled := d.Status.CurrentNumberScheduled
		if numScheduled == 0 {
			return 0, fmt.Errorf("a daemonset expects zero scheduled client pods; cannot validate with no running clients")
		}
		if agreedNumberScheduled == 0 {
			agreedNumberScheduled = numScheduled
		} else if numScheduled != agreedNumberScheduled {
			return 0, fmt.Errorf("daemonsets do not all agree on the number of expected client pods")
		}
	}
```

Firstly, if any daemonset has no scheduled pods, we assume that means the daemonset controller hasn't scheduled pods yet for that daemonset. I encountered this a regularly in testing.

Secondly, it waits until all daemonsets have exactly the same number of pods scheduled., just in case the daemonset controller schedules pods opportunistically instead of all-at-once. I didn't actually encounter this in testing, so I think the k8s scheduler will schedule all pods at once, but I wonder if I might see something different in a very large k8s cluster.

Reply:
Yeah. The routine itself looks for scheduled, then running, then ready.